### PR TITLE
[aot] Added C-API behavior tests

### PR DESCRIPTION
**Pull request by damnkk**
Issue: #

### Brief Summary
Added more test cases.

### CODE REVIEW

File: c_api/tests/c_api_behavior_test.cpp [hunk 1]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+      TiRuntime runtime = ti_create_runtime(arch);
+      for (int i = 0; i < 4; ++i) {
+        allocate_info.usage = TI_MEMORY_USAGE_STORAGE_BIT << i;
+        TiMemory memory = ti_allocate_memory(runtime, &allocate_info);
+        TI_ASSERT(memory != TI_NULL_HANDLE);
+      }
+
+      // Attempt to run out of the memory
+      {
+        allocate_info.size = 1000000000000000000;
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:133]: Don't reuse `allocate_info`, it's hard to understand the process when it's constantly changing. Fill it up in each test block, or just make a utility filler function instead.

File: c_api/tests/c_api_behavior_test.cpp [hunk 2]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+TEST_F(CapiTest, TestBehaviorFreeMemory) {
+  auto inner = [](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      TiRuntime runtime = ti_create_runtime(arch);
+      TiMemoryAllocateInfo *allocate_info = new TiMemoryAllocateInfo;
+      allocate_info->size = 1024;
+      allocate_info->usage = TI_MEMORY_USAGE_STORAGE_BIT;
+      TiMemory memory = ti_allocate_memory(runtime, allocate_info);
+      ti_free_memory(runtime, memory);
+      CHECK_TAICHI_SUCCESS();
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:172]: Move this into a block.

File: c_api/tests/c_api_behavior_test.cpp [hunk 3]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  TiMemoryAllocateInfo allocate_info;
+  allocate_info.size = 1024;
+  allocate_info.usage = TI_MEMORY_USAGE_STORAGE_BIT;
+
+  auto inner = [&](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      TiRuntime runtime = ti_create_runtime(arch);
+      TiMemory memory = ti_allocate_memory(runtime, &allocate_info);
+      ti_map_memory(runtime, memory);
+      CHECK_TAICHI_SUCCESS();
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:211]: Same.

Comment (PENGUINLIONG) [reply c_api/tests/c_api_behavior_test.cpp:211]: Also you should check whether the mapped pointer is non-null; and remember to unmap the memory.

File: c_api/tests/c_api_behavior_test.cpp [hunk 4]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  inner(TI_ARCH_VULKAN);
+}
+
+TEST_F(CapiTest, TestBehaviorUnmapMemory) {
+  TiMemoryAllocateInfo allocate_info;
+  allocate_info.size = 1024;
+  auto inner = [&](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      TiRuntime runtime = ti_create_runtime(arch);
+      for (int i = 0; i < 4; ++i) {
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:244]: Why loop 4?

File: c_api/tests/c_api_behavior_test.cpp [hunk 5]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  TiImageExtent extent;
+  extent.height = 512;
+  extent.width = 512;
+  extent.depth = 1;
+  extent.array_layer_count = 1;
+  TiImageAllocateInfo imageAllocateInfo;
+  imageAllocateInfo.dimension = TI_IMAGE_DIMENSION_2D;
+  imageAllocateInfo.format = TI_FORMAT_RGBA8;
+  imageAllocateInfo.extent = extent;
+  imageAllocateInfo.usage = TI_IMAGE_USAGE_STORAGE_BIT;
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:287]: Check if it sets a `TI_ERROR_ARGUMENT_OUT_OF_RANGE` if usage is `TI_IMAGE_USAGE_MAX_ENUM`.

File: c_api/tests/c_api_behavior_test.cpp [hunk 6]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+TEST_F(CapiTest, TestBehaviorAllocateMemory) {
+  TiError error = TI_ERROR_SUCCESS;
+  TiMemoryAllocateInfo allocate_info;
+  allocate_info.size = 1024;
+  auto inner = [&](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      // Attempt to allocate memory with size of 1024
+      TiRuntime runtime = ti_create_runtime(arch);
+      for (int i = 0; i < 4; ++i) {
+        allocate_info.usage = TI_MEMORY_USAGE_STORAGE_BIT << i;
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:126]: Check if it sets a `TI_ERROR_ARGUMENT_OUT_OF_RANGE` if usage is `TI_MEMORY_USAGE_MAX_ENUM`.

File: c_api/tests/c_api_behavior_test.cpp [hunk 7]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+    }
+  };
+  inner(TI_ARCH_VULKAN);
+}
+
+TEST_F(CapiTest, TestBehaviorAllocateImage) {
+  TiImageExtent extent;
+  extent.height = 512;
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:279]: What if width/height/depth is 0 or a huge number?

File: c_api/tests/c_api_behavior_test.cpp [hunk 8]
```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+      ti_copy_memory_device_to_device(TI_NULL_HANDLE, &dst_memory, &src_memory);
+      CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+
+      // dst memory is null;
+      dst_memory.memory = TI_NULL_HANDLE;
+      ti_copy_memory_device_to_device(runtime, &dst_memory, &src_memory);
+      CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+
+      // src memory is null;
+      src_memory.memory = TI_NULL_HANDLE;
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:452]: Again, remember to release your allocation and runtime.

File: taichi/rhi/opengl/opengl_device.cpp [hunk 9]
```diff
@@ -546,7 +546,6 @@ DeviceAllocation GLDevice::allocate_memory(const AllocParams &params) {
   check_opengl_error("glGenBuffers");
   glBindBuffer(target_hint, buffer);
   check_opengl_error("glBindBuffer");
-
```
Comment (PENGUINLIONG) [inline taichi/rhi/opengl/opengl_device.cpp:549]: Whats this?

**Review by PENGUINLIONG — CHANGES_REQUESTED** [review]
Thank you! Please fix the issues in review comments and lets see if it works.

File: c_api/tests/c_api_behavior_test.cpp (hunk 4 above)
Comment (damnkk) [reply c_api/tests/c_api_behavior_test.cpp:244]: Four kinds of different memory usage bit.

File: c_api/tests/c_api_behavior_test.cpp (hunk 6 above)
Comment (PENGUINLIONG) [reply c_api/tests/c_api_behavior_test.cpp:127]: In this case you only need `TI_MEMORY_USAGE_STORAGE_BIT` checked. That's the only flag required by Taichi Runtime. The others are for the underlying RHI. See the documentation for detail: https://docs.taichi-lang.org/docs/master/taichi_core#bitfield-timemoryusageflags.

For texture flags, check `TI_IMAGE_USAGE_STORAGE_BIT` and `TI_IMAGE_USAGE_SAMPLED_BIT`.

File: c_api/tests/c_api_behavior_test.cpp [hunk 10]
```diff
@@ -113,3 +113,504 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
 
   inner(TI_ARCH_VULKAN);
 }
+
+TEST_F(CapiTest, TestBehaviorAllocateMemory) {
+  TiError error = TI_ERROR_SUCCESS;
+
+  auto inner = [&](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:121]: Is there a particular reason we have to stuff everything in this branch? See the impl of `TestBehaviorGetRuntimeCapabilities`.

File: c_api/tests/c_api_behavior_test.cpp [hunk 11]
```diff
@@ -113,3 +113,504 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  };
+
+  inner(TI_ARCH_VULKAN);
+}
+
+TEST_F(CapiTest, TestBehaviorFreeMemory) {
+  auto inner = [](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      {
+        TiRuntime runtime = ti_create_runtime(arch);
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:173]: destroy

File: c_api/tests/c_api_behavior_test.cpp [hunk 12]
```diff
@@ -113,3 +113,504 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+        allocate_info.size = 1024;
+        allocate_info.usage = TI_MEMORY_USAGE_STORAGE_BIT;
+        TiMemory memory = ti_allocate_memory(runtime, &allocate_info);
+        ti_free_memory(TI_NULL_HANDLE, memory);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+
+      // runtime is not null and allocate_info is null
+      {
+        TiRuntime runtime = ti_create_runtime(arch);
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:200]: destroy

File: c_api/tests/c_api_behavior_test.cpp [hunk 13]
```diff
@@ -113,3 +113,504 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+
+      // runtime & memory are both null
+      {
+        ti_map_memory(TI_NULL_HANDLE, TI_NULL_HANDLE);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+
+      // runtime is null, memory is valid
+      {
+        TiRuntime runtime = ti_create_runtime(arch);
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:234]: please ensure everything you create or allocate is destroyed. we want to keep the CI stable. Or just use something that automatically does the destroy in its destructor like `ti::Runtime`, if the object itself (`TiRuntime` in this case) is not the tested thing.

**Review by PENGUINLIONG — CHANGES_REQUESTED** [review]

File: c_api/tests/c_api_behavior_test.cpp [hunk 14]
```diff
@@ -23,6 +23,8 @@ TEST_F(CapiTest, TestBehaviorCreateRuntime) {
     TiRuntime runtime = ti_create_runtime(arch);
     TI_ASSERT(runtime == TI_NULL_HANDLE);
     CHECK_TAICHI_ERROR_IS(TI_ERROR_NOT_SUPPORTED);
+    ti_destroy_runtime(runtime);
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:26]: well you dont have to destroy if its already failing... it is an expected behavior to recover everything when a create call fails.

File: taichi/rhi/opengl/opengl_device.cpp (hunk 9 above)
Comment (PENGUINLIONG) [reply taichi/rhi/opengl/opengl_device.cpp:549]: If this is an unintentional empty line, please remove it

File: tests/cpp/aot/python_scripts/graph_aot_test_vulkan.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
+import argparse
```
Comment (PENGUINLIONG) [inline tests/cpp/aot/python_scripts/graph_aot_test_vulkan.py:1]: Can we reuse some other AOT script so we don't have many similar scripts?

File: c_api/tests/c_api_behavior_test.cpp [hunk 16]
```diff
@@ -109,7 +110,554 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+      TiRuntime runtime = ti_create_runtime(TI_ARCH_VULKAN);
+      ti_unmap_memory(runtime, TI_NULL_HANDLE);
+      CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      ti_destroy_runtime(runtime);
+    }
+  };
   inner(TI_ARCH_VULKAN);
 }
+
+TiImageAllocateInfo getImageAllocateInfo() {
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:300]: Please follow Taichi's naming convention.

File: c_api/tests/c_api_behavior_test.cpp [hunk 17]
```diff
@@ -109,7 +110,554 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  }
+
+  TiRuntime runtime = ti_create_runtime(arch);
+  TiAotModule module = ti_load_aot_module(runtime, module_path.c_str());
+  TiComputeGraph Cgraph = ti_get_aot_module_compute_graph(module, "run_graph");
+  CHECK_TAICHI_SUCCESS();
+  TI_ASSERT(Cgraph != TI_NULL_HANDLE);
+
+  // Attemp to get compute graph with null module.
+  Cgraph = ti_get_aot_module_compute_graph(TI_NULL_HANDLE, "run_graph");
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:644]: naming convention

File: taichi/rhi/opengl/opengl_device.cpp (hunk 9 above)
Comment (PENGUINLIONG) [reply taichi/rhi/opengl/opengl_device.cpp:549]: Again. Please remove it if it's an unintentional change.

File: c_api/tests/c_api_behavior_test.cpp [hunk 18]
```diff
@@ -109,7 +110,554 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  // Attemp to get compute graph with invalid name.
+  Cgraph = ti_get_aot_module_compute_graph(module, "#$#%*(");
+  CHECK_TAICHI_ERROR_IS(TI_ERROR_NAME_NOT_FOUND);
+  TI_ASSERT(Cgraph == TI_NULL_HANDLE);
+
+  ti_destroy_runtime(runtime);
+}
+
+TEST_F(CapiTest, TestBehaviorGetCgraphVulkan) {
+  test_behavir_get_cgraph_impl(TI_ARCH_VULKAN);
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:662]: typo

File: c_api/tests/c_api_behavior_test.cpp [hunk 19]
```diff
@@ -109,7 +110,554 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
...
+  CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+}
+
+TEST_F(CapiTest, TestBehaviorDestroyAotModuleVulkan) {
+  test_behavior_destroy_aot_module_impl(TI_ARCH_VULKAN);
+}
+
+void test_behavir_get_cgraph_impl(TiArch arch) {
+  const auto folder_dir = getenv("TAICHI_AOT_FOLDER_PATH");
+  const std::string module_path = folder_dir + std::string("/module.tcm");
```
Comment (PENGUINLIONG) [inline c_api/tests/c_api_behavior_test.cpp:631]: You can simply load it from `folder_dir` then you no longer need to change `graph_aot_test.py`.

**Review by PENGUINLIONG — CHANGES_REQUESTED** [review]

**Pull request merged**